lib/workflow/postgres/postgres.go

```diff
@@ -0,0 +1,726 @@
+	"github.com/sirupsen/logrus"
+
+	"github.com/chef/automate/lib/workflow"
+	"github.com/chef/automate/lib/workflow/backend"
+)
+
+const (
+	enqueueTaskQuery   = `SELECT enqueue_task($1, $2, $3, $4, $5)`
+	dequeueTaskQuery   = `SELECT * FROM dequeue_task($1)`
+	completeTaskQuery  = `SELECT complete_task($1::bigint, $2::task_status, $3::text, $4)`
+	getTaskResultQuery = `SELECT task_name, parameters, status, error, result from tasks_results WHERE id = $1`
+
+	enqueueWorkflowQuery  = `SELECT enqueue_workflow($1, $2, $3)`
+	dequeueWorkflowQuery  = `SELECT * FROM dequeue_workflow(VARIADIC $1)`
+	completeWorkflowQuery = `SELECT complete_workflow($1)`
+	continueWorkflowQuery = `SELECT continue_workflow($1, $2, $3, $4, $5)`
+	abandonWorkflowQuery  = `SELECT abandon_workflow($1, $2, $3)`
+
+	listRecurringWorkflowsQuery = `
+		WITH res AS
+			(SELECT DISTINCT ON (name, workflow_name) * from workflow_results
+			ORDER BY name, workflow_name, end_at DESC)
+		SELECT s.id, enabled, s.name, s.workflow_name, s.parameters, recurrence,
+			next_run_at, start_at last_start, end_at last_end
+			FROM recurring_workflow_schedules s LEFT JOIN res
+			ON s.name = res.name AND s.workflow_name = res.workflow_name;
+		`
+
+	getNextRecurringWorkflowQuery = `
+        SELECT id, enabled, name, workflow_name, parameters, recurrence, next_run_at
+        FROM recurring_workflow_schedules
+        WHERE enabled = TRUE
+        ORDER BY next_run_at LIMIT 1
+        `
+	getDueRecurringWorkflowQuery = `
+        SELECT id, enabled, name, workflow_name, parameters, recurrence, next_run_at
+        FROM recurring_workflow_schedules
+        WHERE next_run_at < NOW() AND enabled = TRUE
+        ORDER BY next_run_at
+        FOR UPDATE SKIP LOCKED LIMIT 1
+        `
+	updateRecurringWorkflowQuery = `
+		UPDATE recurring_workflow_schedules 
+		SET next_run_at = $2, 
+		last_enqueued_at = $3,
+		enabled = $4
+		WHERE id = $1
+        `
+
+	updateSlowRecurringWorkflowQuery = `
+		UPDATE recurring_workflow_schedules 
+		SET next_run_at = $2,
+		enabled = $3
+		WHERE id = $1
+        `
+)
+
+type PostgresBackend struct {
+	connURI string
+	db      *sql.DB
+}
+
+type PostgresTaskCompleter struct {
+	// tid is the Task's id in our postgresql database.  We need
+	// this to complete the correct task.
+	tid int64
+
+	_txLock sync.Mutex
+	// _tx is the transaction that is holding our dequeued task.
+	_tx *sql.Tx
+	// ctx is the context that the transaction was started with.
+	ctx context.Context
+	// canceling this will abort the transaction
+	cancel context.CancelFunc
+}
+
+type PostgresWorkflowCompleter struct {
+	cancel context.CancelFunc
+	// wid is the WorkflowInstance id in our postgresql database.  We need
+	// this to complete the correct workflow
+	wid int64
+	// eid is the WorkflowEvent id
+	eid int64
+
+	// tx is the transaction that is holding our dequeued workflow.
+	tx *sql.Tx
+	// ctx is the context that the transaction was started with.
+	ctx context.Context
```

Comment:
Currently we store the ctx, the cancellation func, and the tx.  Unfortunately, because of bugs in the pq library, we don't think it is safe to refer on simply cancelling this context and thus we should probably be handling rollback ourselves.